src/buckets.spec.ts

```diff
@@ -0,0 +1,200 @@
+    // @todo: Should we rename itemsList to just be items?
+    expect(res.item?.itemsList).to.have.length(0)
+  })
+
+  it('should push data from filesystem on node', async function () {
+    if (isBrowser) return this.skip()
+    const pth = path.join(__dirname, '..', 'testdata')
+    fileSize = fs.statSync(path.join(pth, 'file1.jpg')).size
+    let stream = fs.createReadStream(path.join(pth, 'file1.jpg'))
+    const rootKey = buck.root?.key || ''
+    let length = 0
+
+    // Bucket path
+    const res = await client.pushPath(rootKey, 'dir1/file1.jpg', stream, ctx, {
+      progress: (num) => (length = num || 0),
+    })
+    expect(length).to.equal(fileSize)
+    expect(res.path).to.not.be.undefined
+    expect(res.root).to.not.be.undefined
+
+    // Nested bucket path
+    stream = fs.createReadStream(path.join(pth, 'file2.jpg'))
+    const { root } = await client.pushPath(rootKey, 'path/to/file2.jpg', stream, ctx)
+    expect(root).to.not.be.undefined
+
+    // Root dir
+    const rep = await client.listPath(rootKey, '', ctx)
+    expect(rep.item?.isdir).to.be.true
+    expect(rep.item?.itemsList).to.have.length(2)
+  })
+
+  it('should push data from file API in browser', async function () {
+    if (isNode) return this.skip()
+    const parts = [
+      new Blob(['you construct a file...'], { type: 'text/plain' }),
+      ' Same way as you do with blob',
+      new Uint16Array([33]),
+    ]
+    // Construct a file
+    const file = new File(parts, 'file1.txt')
+    const rootKey = buck.root?.key || ''
+    let length = 0
+
+    // Bucket path
+    const res = await client.pushPath(rootKey, 'dir1/file1.jpg', file, ctx, {
+      progress: (num) => (length = num || 0),
+    })
+    expect(length).to.equal(54)
+    expect(res.path).to.not.be.undefined
+    expect(res.root).to.not.be.undefined
+
+    // Nested bucket path
+    // @note: We're reusing file here...
+    const { root } = await client.pushPath(rootKey, 'path/to/file2.jpg', file, ctx)
+    expect(root).to.not.be.undefined
+
+    // Root dir
+    const rep = await client.listPath(rootKey, '', ctx)
+    expect(rep.item?.isdir).to.be.true
+    expect(rep.item?.itemsList).to.have.length(2)
+  })
+
+  it('should list (nested) files within a bucket', async () => {
+    const rootKey = buck.root?.key || ''
+
+    // Nested dir
+    let rep = await client.listPath(rootKey, 'dir1', ctx)
+    expect(rep.item?.isdir).to.be.true
+    expect(rep.item?.itemsList).to.have.length(1)
+
+    // File
+    rep = await client.listPath(rootKey, 'dir1/file1.jpg', ctx)
+    expect(rep.item?.path.endsWith('file1.jpg')).to.be.true
+    expect(rep.item?.isdir).to.be.false
+  })
+
+  it('should pull files by path and write to file on node', async function () {
+    if (isBrowser) return this.skip()
+    // Bucket path
+    const rootKey = buck.root?.key || ''
+    let length = 0
+
+    // Bucket path
+    const chunks = client.pullPath(rootKey, 'dir1/file1.jpg', ctx, {
```

Comment:
This returns an AsyncIterableIterator of UInt8Arrays (Buffers), which can easily be fed into a write stream, or file write object!